client/beefy/src/worker.rs

```diff
@@ -529,18 +559,41 @@ where
 		let best_grandpa = *self.best_grandpa_block_header.number();
 		let _ph = PhantomData::<B>::default();
 
-		fn to_process_for<B: Block, T>(
-			pending: &mut BTreeMap<NumberFor<B>, T>,
+		fn to_process_for<B: Block, T, const U: u32>(
+			pending: &mut BoundedBTreeMap<NumberFor<B>, T, ConstU32<U>>,
 			(start, end): (NumberFor<B>, NumberFor<B>),
 			_: PhantomData<B>,
-		) -> BTreeMap<NumberFor<B>, T> {
-			// These are still pending.
-			let still_pending = pending.split_off(&end.saturating_add(1u32.into()));
-			// These can be processed.
-			let to_handle = pending.split_off(&start);
-			// The rest can be dropped.
-			*pending = still_pending;
-			// Return ones to process.
+		) -> BTreeMap<NumberFor<B>, T>
+		where
+			u64: From<NumberFor<B>>,
+			<<B as BlockT>::Header as HeaderT>::Number: From<u64>,
+		{
+			let mut to_handle = BTreeMap::new();
+
+			let mut still_pending = BTreeMap::new();
+
+			let end: u64 = end.into();
+			let start: u64 = start.into();
+			let still_pending_range = end.saturating_add(1u32.into())..U.into();
+
+			for i in still_pending_range {
+				let value_to_be_removed = pending.remove(&i.into());
+				if value_to_be_removed.is_some() {
+					still_pending.insert(i.into(), value_to_be_removed.unwrap());
+				}
+			}
+
+			let to_handle_range = start..=end;
+
+			for i in to_handle_range {
+				let value_to_be_removed = pending.remove(&i.into());
+				if value_to_be_removed.is_some() {
+					to_handle.insert(i.into(), value_to_be_removed.unwrap());
+				}
+			}
+
+			*pending = BoundedBTreeMap::checked_from(still_pending).expect("Should not fail");
+
```

Comment:
I was actually having issue using the `split_off` logic due to the fact that we are now using a `BoundedBTreeMap`

Reply:
Maybe `BoundedBTreeMap` is not the right tool then. Existing `BoundedBTreeMap` has forced you into costly implementations for both the `enqueue` and `process` operations:
- enqueue: now removes vec from tree, mutates vec, then inserts it back -> two `O(logN)` ops instead of a simple `O(1)` of just mutating value (vec) of some constant key (block num);
- process: now iterating and _copying_ data in memory from one treemap to another on each step, when before `split_off` would just do some cheap pointer assignments and not move any data in memory.

I suggest either:
1. enhance `BoundedBTreeMap` with new methods to do what we need, or,
2. create your own local very lightweight `BTreeMap` wrapper object that has only two or three that we actually need, you can then just do the logic just as before with an extra manually check `BTreeMap::len()` in those methods.

Reply:
So, the issue is when trying to transform a `BTreeMap` to a `BoundedBTreeMap` in the `to_process_for` method, i get this error:

  *pending = still_pending;
    |             --------   ^^^^^^^^^^^^^ expected struct `BoundedBTreeMap`, found struct `BTreeMap`
    |             |
    |             expected due to the type of this binding
    |
    = note: expected struct `BoundedBTreeMap<<<B as BlockT>::Header as HeaderT>::Number, T, ConstU32<U>>`
               found struct `BTreeMap<<<B as BlockT>::Header as HeaderT>::Number, T>`

Reply:
> the issue is when trying to transform a `BTreeMap` to a `BoundedBTreeMap`

Why do you want to do this?

I was suggesting to create your own (private) wrapper:

```rust
struct MyBoundedBTreeMap<K, V> {
    max_len: usize,
    inner: BtreeMap<K, V>,
}
impl<K, V> MyBoundedBTreeMap<K, V> {
    fn new(max_len: usize) -> Self { BoundedBTreeMap { max_len, inner: BtreeMap::new() } }
    fn split_off(&mut self, key: &K) -> Self {
        let other_inner = self.inner.split_off(key);
        BoundedBTreeMap { max_len: self.max_len, inner: other_inner }
    }
    fn get_mut(&mut self, key: &K) -> Option<&mut V> { self.inner.get_mut(key) }
    fn try_insert(&mut self, key: K, value: V) -> Result<()> {
        if self.inner.len() >= self.max_len { return Err() }
        self.insert(K, V);
    }
}
```

And these ^ methods are all you need to use in the voter.